File: tests/perl/src/T015_FieldSpecification.pl
#!/usr/bin/env perl

use strict;
use Test;
use libgmsec_perl;


sub Test_FieldSpecification
{
	my ($test) = @_;

	$test->setDescription("Test FieldSpecification");

	test_field_specification($test);
}


sub test_field_specification
{
	libgmsec_perl::logInfo("Test FieldSpecification");

	my ($test) = @_;

	test_field_specification_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_CURRENT, $libgmsec_perl::Specification::SchemaLevel_LEVEL_0);
	test_field_specification_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_CURRENT, $libgmsec_perl::Specification::SchemaLevel_LEVEL_1);
	test_field_specification_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_CURRENT, $libgmsec_perl::Specification::SchemaLevel_LEVEL_2);

	test_field_specification_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_2019_00, $libgmsec_perl::Specification::SchemaLevel_LEVEL_0);
	test_field_specification_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_2019_00, $libgmsec_perl::Specification::SchemaLevel_LEVEL_1);
	test_field_specification_aux($test, $libgmsec_perl::GMSEC_MSG_SPEC_2019_00, $libgmsec_perl::Specification::SchemaLevel_LEVEL_2);
}


sub test_field_specification_aux
{
	my ($test, $specVersion, $schemaLevel) = @_;

	my $config = libgmsec_perl::Config->new();
	$config->addValue("gmsec-specification-version", "$specVersion");
	$config->addValue("gmsec-schema-level", "$schemaLevel");

	my $spec = libgmsec_perl::Specification->new($config);

	my $msgSpecs = $spec->getMessageSpecifications();

	for (my $i = 0; $i < $msgSpecs->size(); $i++)
	{
		my $msgSpec = $msgSpecs->get($i);

		my $fieldSpecs = $msgSpec->getFieldSpecifications();

		$test->check("Expected to have at least 1 (one) field specification", $fieldSpecs->size() > 0);

		for (my $j = 0; $j < $fieldSpecs->size(); $j++)
		{
			my $fieldSpec = $fieldSpecs->get($j);

			$test->check("FieldSpecification name is empty", defined $fieldSpec->getName());
			$test->check("FieldSpecification type is empty", defined $fieldSpec->getType());
			$test->check("FieldSpecification mode is empty", defined $fieldSpec->getMode());
			$test->check("FieldSpecification classification is empty", defined $fieldSpec->getClassification());
			$test->check("FieldSpecification value is empty", defined $fieldSpec->getValue());
			$test->check("FieldSpecification description is empty", defined $fieldSpec->getDescription());
		}
	}
}


Test::driver(\&Test_FieldSpecification);

File: tests/python3/src/T015_FieldSpecification.py
#!/usr/bin/env python3

import libgmsec_python3 as lp
from Test import Test


class Test_FieldSpecification(Test):

    def test(self):
        self.test_field_specification()


    def test_field_specification(self):
        lp.log_info("Test FieldSpecification")

        self.test_field_specification_aux(lp.GMSEC_MSG_SPEC_CURRENT, lp.Specification.SchemaLevel_LEVEL_0)
        self.test_field_specification_aux(lp.GMSEC_MSG_SPEC_CURRENT, lp.Specification.SchemaLevel_LEVEL_1)
        self.test_field_specification_aux(lp.GMSEC_MSG_SPEC_CURRENT, lp.Specification.SchemaLevel_LEVEL_2)

        self.test_field_specification_aux(lp.GMSEC_MSG_SPEC_2019_00, lp.Specification.SchemaLevel_LEVEL_0)
        self.test_field_specification_aux(lp.GMSEC_MSG_SPEC_2019_00, lp.Specification.SchemaLevel_LEVEL_1)
        self.test_field_specification_aux(lp.GMSEC_MSG_SPEC_2019_00, lp.Specification.SchemaLevel_LEVEL_2)


    def test_field_specification_aux(self, specVersion, schemaLevel):
        config = lp.Config()

        config.add_value("gmsec-specification-version", "" + str(specVersion))
        config.add_value("gmsec-schema-level", "" + str(schemaLevel))

        spec = lp.Specification(config)

        msgSpecs = spec.get_message_specifications()

        for msgSpec in msgSpecs:
            fieldSpecs = msgSpec.get_field_specifications()

            self.check("Expected to have at least 1 (one) field specification", fieldSpecs.size() > 0)

            for fieldSpec in fieldSpecs:
                self.check("FieldSpecification name is None", fieldSpec.get_name() != None)
                self.check("FieldSpecification type is None", fieldSpec.get_type() != None)
                self.check("FieldSpecification mode is None", fieldSpec.get_mode() != None)
                self.check("FieldSpecification classification is None", fieldSpec.get_classification() != None)
                self.check("FieldSpecification value is None", fieldSpec.get_value() != None)
                self.check("FieldSpecification description is None", fieldSpec.get_description() != None)


Test.run('Test FieldSpecification', Test_FieldSpecification())
